supabase/tests/refresh_tokens.test.sql

```diff
@@ -0,0 +1,108 @@
+create function tests.startup_refresh_tokens_as_alice()
+returns setof text as $$
+begin
+
+  -- Note that seed.sql installs fitures into auth.users (alice, bob, carol)
+  -- as well as user_grants (to aliceCo/, bobCo/, carolCo/).
+  set request.jwt.claim.sub to '11111111-1111-1111-1111-111111111111';
+
+end;
+$$ language plpgsql;
+
+create function tests.test_create_refresh_token()
+returns setof text as $$
+declare
+  response json;
+begin
+  delete from refresh_tokens;
+  select create_refresh_token(false, '1 day', 'test detail') into response;
+
+  return query select results_eq(
+    $i$ select multi_use, valid_for, detail, uses, user_id from refresh_tokens $i$,
+    $i$ values (false, interval '1 day', 'test detail', 0, '11111111-1111-1111-1111-111111111111'::uuid) $i$,
+    'initial refresh_token created'
+  );
+
+  return query select ok(response::jsonb ? 'id', 'refresh_token response has id');
+  return query select ok(response::jsonb ? 'secret', 'refresh_token response has secret');
+end;
+$$ language plpgsql;
+
+
+create function tests.test_generate_access_token()
+returns setof text as $$
+declare
+  rt refresh_tokens;
+  new_rt refresh_tokens;
+  rt_response jsonb;
+  response json;
+begin
+  delete from refresh_tokens;
+
+  select create_refresh_token(true, '1 day', 'test detail') into rt_response;
+  select * into rt from refresh_tokens;
+
+  -- Multi-use refresh_token used for this one, so expect only an access_token
+  -- in response. Expect an update to `updated_at` and `uses` of the
+  -- refresh_token
+  select generate_access_token((rt_response->>'id')::flowid, rt_response->>'secret') into response;
+
+  return query select ok(response::jsonb ? 'access_token', 'generate_access_token response has access_token');
+  return query select ok(not (response::jsonb ? 'refresh_token'), 'generate_access_token response does not have refresh_token (multi-use)');
+
+  select * into new_rt from refresh_tokens;
+  return query select is(new_rt.uses, 1, 'refresh_tokens uses bumped');
+  return query select is(new_rt.hash, rt.hash, 'refresh_token hash unchanged (multi-use)');
+
+  -- Unfortunately we can't run this test because `now()` refers to the
```

Comment:
you can use clock_timestamp() instead of now() to get the actual time, even inside of a transaction. IMO it would be reasonable to use clock_timestamp() in the actual function bodies in order to make the tests work (we don't care that it's wall-time vs transaction-start-time in practice).